File: 0733-flood-fill/0733-flood-fill.cpp
class Solution {
public:
    void dfs(int i,int j,vector<vector<int>>& image,int color,int sameColor){
        if(i<0 or j<0 or i>=image.size() or j>=image[0].size() or image[i][j]!=sameColor or image[i][j]==color)
            return;
        image[i][j] = color;
        dfs(i+1,j,image,color,sameColor);
        dfs(i-1,j,image,color,sameColor);
        dfs(i,j+1,image,color,sameColor);
        dfs(i,j-1,image,color,sameColor);
    }
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        int sameColor = image[sr][sc];
        dfs(sr,sc,image,color,sameColor);
        return image;
    }
};

File: 0733-flood-fill/0733-flood-fill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        DIR = [0,1,0,-1,0]
        old_col, m, n = image[sr][sc], len(image), len(image[0])
        if old_col != color:
            q = collections.deque([(sr,sc)])
            while q:
                r,c = q.popleft()
                image[r][c] = color
                for i in range(4):
                    nr = r + DIR[i]
                    nc = c + DIR[i+1]
                    if 0<= nr < m and 0<= nc <n and image[nr][nc] == old_col:
                        q.append((nr,nc))
        return image